### ISSUE TITLE
streams

### ISSUE THREAD

**Issue by jonathanong** [coderhaoxin/pg-then#7]
i'm going to add a `.queryStream()` method that wraps around https://github.com/brianc/node-pg-query-stream. any objections?

**Comment by jonathanong** [comment]
hmmm might seem harder that i thought..

**Comment by jonathanong** [comment]
```js
.stream = (text, value, options) => {
  const query = new QueryStream(text, value, options)
  const stream = db._client.query(query)

  stream.on('close', onclose)
  stream.on('error', onerror)
  stream.on('end', onend)

  return stream

  function onerror (err) {
    if (stream.listenerCount('error') === 1) console.error(err.stack)
    cleanup()
  }

  function onend () {
    cleanup()
  }

  function onclose () {
    cleanup()
  }

  function cleanup () {
    stream.removeListener('close', onclose)
    stream.removeListener('error', onerror)
    stream.removeListener('end', onend)
    db._done()
  }
}
```

will make a PR later

**Comment by jonathanong** [comment]
tried it locally and it's failing for queries that return 0 results. HMMM

**Comment by coderhaoxin** [comment]
https://github.com/coderhaoxin/pg-then/pull/11

**Comment by vitaly-t** [comment]
@jonathanong it is indeed harder than you would initially think....

For one thing, you really would want to report the number of records streamed. But when I asked BrianC about this: https://github.com/brianc/node-pg-query-stream/issues/12, he was of no help.

So, when eventually I wrote this for [pg-promise](https://github.com/vitaly-t/pg-promise), I had to hook into his interface and provide my own counter for the records being processed:

```js
stream = ctx.db.client.query(qs);
        fetch = stream._fetch;
        stream._fetch = function (size, func) {
            fetch.call(stream, size, function (err, rows) {
                nRows += err ? 0 : rows.length;
                return func(err, rows);
            });
        };
```

It all eventually worked just fine, with the test coverage to prove that. See also the [streaming example](https://github.com/vitaly-t/pg-promise/wiki/Learn-by-Example#streaming).
